### ISSUE TITLE
[BUG] DALI result is worse than normal pytorch data loader

### ISSUE THREAD

**Issue by agemagician** [NVIDIA/NeMo#2853]
Hello,

When running the same training or even the same validation script without DALI, we get the expected WER.
However, when we enable DALI using the same configuration, we get much worse WER.

I have created a Google Colab example that replicates this issue, which can be found here:
https://colab.research.google.com/drive/1Rz42EeQVDHhTso3kWvBp1wQc6x1FxXgY?usp=sharing

To make the debugging more straightforward, I have disabled the training on the example.
So simply, it performs testing with DALI and without DALI.

This is related to issue #2817 , which was fixed by @titu1994 .

**Comment by agemagician** [comment]
I have checked the output of the data loader for both DALI and PyTorch and the results are totally different for the input audio and input length, but the same for output text and output length.

You can check these results from this Google Colab example:
https://colab.research.google.com/drive/1Bk7fUBnTuIvC7JvDPau0lsuBjCmeOYEx?usp=sharing

It seems there is a bug on the DALI data loader that produces results totally different from the PyTorch data loader.

**Comment by agemagician** [comment]
Besides the 1st post example which uses conformed model, I have also tested QuartzNet model and it have the same issue:
https://colab.research.google.com/drive/1eQe7EjLwjGsO7Dktaj2p9uvb91YYpBQb?usp=sharing

It confirms the same issue.

**Comment by jantonguirao** [comment]
Thank you @agemagician for reporting this and for the easy-to-follow reproduction steps!
I can confirm there is a bug in the usage of DALI inside NeMo.

- Just to clarify: About the output of the data loader being different for the DALI implementation, that is actually expected. This is because the DALI implementation fuses the data loading and the feature calculation (mel spectrogram) into a single step. So when you use DALI for loading, the data loader provides you with a Mel spectrogram directly. The PyTorch based implementation produces the raw audio, and there's a preprocessing step which follows. 

- The bug was a wrong calculation of the window stride, where the window size (not stride) was used instead. https://github.com/NVIDIA/NeMo/pull/2858 should fix this (I tested it, and got test_wer=0.036 after the fix)

**Comment by jantonguirao** [comment]
There might be small differences in each step of the data processing pipeline. Because of that, it might happen that the network is more accurate when using the same preprocessing implementation for both training and inference.

That being said, I am not sure if that kind of error can be attributed to such numerical error and small differences in implementation. I will run some experiments, and find out if there are any other bugs in the code.

Meanwhile, @titu1994 do you think that the loss and WER differences mentioned above can be attributed to small errors between implementations, or is it a significant error that should be investigated?

**Comment by titu1994** [comment]
A wer difference of 0.5% is quite high especially when when done on the clean data subset, indicating that there is mismatch in preprocessing which culminated in mismatched alignments. 

I'd need to see individual samples to make out, but for simple debugging id try to find a single sample that produces different results than try to compare a batch of sample scores. 

Conformer is very sensitive to it's input due to the self attention in it's design, I wonder if we can do an abs(dali_spec-pt_spec) to find source of issue. 

Another thing is Dali pipeline builds the either at compile time, but during test we generally switch off dither. Is there a way to allow dither only when in training mode ?

**Comment by jantonguirao** [comment]
I've found another source of error, it was due to the default value user as a floor for the logarithm calculation. I've submitted another PR (https://github.com/NVIDIA/NeMo/pull/2869) that hopefully will bring the WER to a reasonable value.

I've run some experiments and calculated the absolute difference between the data provided by DALI and the one provided by the torch based implementation. It looks quite similar now:
![image](https://user-images.githubusercontent.com/3891217/134339833-79c5a223-d698-4afd-92cd-332ebea00e05.png)
![image](https://user-images.githubusercontent.com/3891217/134339944-b7d5e77d-f141-4c9e-be30-c1cda3fcbf58.png)

**Comment by agemagician** [comment]
Thanks a lot, @jantonguirao, for your effort; I have re-tested it, and now it is almost identical :

------------------------------

Conformer Large:
Pytorch Data Loader: {'test_loss': 3.161954402923584, 'test_wer': 0.02493533119559288}
DALI: {'test_loss': 3.1656789779663086, 'test_wer': 0.02499239146709442}

The current difference in loss is "0.00372457504" and WER "scaled from 0 to 1 to 0 - 100" is "0.00570602715".

------------------------------

Conformer Small:
Pytorch Data Loader: {'test_loss': 4.655806541442871, 'test_wer': 0.040360622107982635}
DALI: {'test_loss': 4.655491352081299, 'test_wer': 0.04039866104722023}

The current difference in loss is "0.00031518936" and WER "scaled from 0 to 1 to 0 - 100" is "0.00380389392".

------------------------------

I have also tested it during training and the loss is the same between PyTorch Data Loader and DALI.

------------------------------

For me personally, I am very satisfied with this results and this fix the DALI pipeline

**Comment by jantonguirao** [comment]
Glad to see it works as expected now!
@titu1994 I think the current error between DALI and torch implementations is small enough now, but let me know if you think we should investigate this further

**Comment by titu1994** [comment]
This is fantastic ! Amazing resolution @jantonguirao ! I'll review and merge asap

**Comment by titu1994** [comment]
That were difference is well within the margin of error between two implementations. This is good to merge
